### PR TITLE
Implement Houdini worklets

### LINKED ISSUE #16206: Implement CSS Houdini Worklets
Draft spec: https://drafts.css-houdini.org/worklets/
Houdini CSS issue for servo: #13557 

cc @tschneidereit @metajack @jdm @SimonSapin

### PR DESCRIPTION
**Pull request by asajeffrey**
This PR implements the current draft Houdini Worklets specification (https://drafts.css-houdini.org/worklets/).

The implementation is intended to provide a responsive environment for worklet execution, and in particular to ensure that the primary worklet executor does not garbage collect, and does not block loading module code. The implementation does this by providing a thread pool, and performing GC and module loading in a backup thread, not in the primary thread.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #16206 
- [x] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/16814)

### CODE REVIEW

**Comment by asajeffrey** [comment]
cc @tschneidereit @metajack

**Comment by metajack** [comment]
cc @bfgeek

**Comment by asajeffrey** [comment]
Some things that should probably be fixed before this lands:
 * the methods in `globalscope` which use downcasting are missing some `WorkletGlobalScope` cases.
 * threads should be named.
 * add time reporting code.
 * think about how to configure SM for worklet threads (in particular how to discourage GC).

The code is ready for review as-is though.

**Comment by asajeffrey** [comment]
Hi @bfgeek! This is implementation of worklets in Servo, using a thread pool to ensure worklet execution isn't blocked by GC or by module loading. Mostly the spec was pretty straightforward to implement, you'll be pleased to hear!

**Comment by asajeffrey** [comment]
The spec issues raised by implementing worklets (and starting on paint worklets):

* Speculative evalution (https://github.com/w3c/css-houdini-drafts/issues/406)
* Fetching and invoking a worklet script which reports an exception (https://github.com/w3c/css-houdini-drafts/issues/407)

**Comment by bfgeek** [comment]
cc/ @nhiroki who works on our implementation. 

That's great to see! Thanks for filing spec issues.

File: components/script/dom/webidls/Window.webidl
Comment (jdm) [inline components/script/dom/webidls/Window.webidl]: What if we made a separate DOM interface with a constructor so we don't need to include test-only fields in Window?

File: components/script/dom/workletglobalscope.rs
Comment (jdm) [inline components/script/dom/workletglobalscope.rs]: Let's use channel instead of chan in these comments.

File: components/script/dom/worklet.rs
Comment (jdm) [inline components/script/dom/worklet.rs]: Any reason not to collapse some of these import statements from the same modules?

File: components/script/dom/worklet.rs
Comment (jdm) [inline components/script/dom/worklet.rs]: We should use `reflect_dom_object` instead.

File: components/script/dom/worklet.rs
Comment (jdm) [inline components/script/dom/worklet.rs]: ?

File: components/script/dom/worklet.rs
Comment (jdm) [inline components/script/dom/worklet.rs]: I would prefer using `#[ignore_heap_size_of = "Can't measure uuid"]` instead.

File: components/script/dom/worklet.rs
Comment (jdm) [inline components/script/dom/worklet.rs]: Why AcqRel instead of SeqCst?

File: components/script/dom/worklet.rs
Comment (jdm) [inline components/script/dom/worklet.rs]: This happens automatically when events run in the script thread. Did you mean something more specific?

File: components/script/dom/worklet.rs
Comment (jdm) [inline components/script/dom/worklet.rs]: nit: g

**Review by jdm — CHANGES_REQUESTED** [review]

**Comment by jdm** [comment]
As usual, I found these changes to be a pleasant and straightforward read. Good job!

File: components/script/dom/worklet.rs
Comment (jdm) [inline components/script/dom/worklet.rs]: We should be using CORS mode here, per https://html.spec.whatwg.org/multipage/webappapis.html#fetch-a-single-module-script. We'll need to include a real origin, too.

File: components/script/dom/webidls/Window.webidl
Comment (asajeffrey) [inline components/script/dom/webidls/Window.webidl]: Will do.

File: components/script/dom/workletglobalscope.rs
Comment (asajeffrey) [inline components/script/dom/workletglobalscope.rs]: Will do.

File: components/script/dom/worklet.rs
Comment (asajeffrey) [inline components/script/dom/worklet.rs]: Minimizing the chance of annoying merge conflicts.

File: components/script/dom/worklet.rs
Comment (asajeffrey) [inline components/script/dom/worklet.rs]: I don't think you can use `reflect_dom_object` to create globals can you?

File: components/script/dom/worklet.rs
Comment (asajeffrey) [inline components/script/dom/worklet.rs]: This is dead code, just in for future-proofing. I can get rid of it.

File: components/script/dom/worklet.rs
Comment (asajeffrey) [inline components/script/dom/worklet.rs]: Will do.

File: components/script/dom/worklet.rs
Comment (asajeffrey) [inline components/script/dom/worklet.rs]: Using the weakest ordering that does the job. In this case it doesn't matter a huge amount, this isn't hot code.

File: components/script/dom/worklet.rs
Comment (asajeffrey) [inline components/script/dom/worklet.rs]: Oops.

File: components/script/dom/worklet.rs
Comment (asajeffrey) [inline components/script/dom/worklet.rs]: CORS mode isn't a problem, but I'm not sure which origin to use here. One snag is that this is being done in a different thread, so we don't have access to the `MutableOrigin` of the window.

File: components/script/dom/worklet.rs
Comment (asajeffrey) [inline components/script/dom/worklet.rs]: At this point, we have loaded a worklet, but not actually run it (e.g. to get a paint worklet to do its painting).

**Review by asajeffrey — COMMENTED** [review]
Thanks for the review!

File: components/script/dom/worklet.rs
Comment (jdm) [inline components/script/dom/worklet.rs]: WorkletBinding isn't a global.

File: components/script/dom/worklet.rs
Comment (jdm) [inline components/script/dom/worklet.rs]: Send an ImmutableOrigin from the window as part of the initiation?

File: components/script/dom/worklet.rs
Comment (asajeffrey) [inline components/script/dom/worklet.rs]: Oh good point, we're in `Worklet` not `WorkletGlobalScope`.

**Review by jdm — APPROVED** [review]

**Comment by jdm** [comment]
Squash and r=me.

**Comment by jdm** [comment]
```
  ▶ Unexpected subtest result in /_mozilla/mozilla/interfaces.html:
  │ FAIL [expected PASS] Interfaces exposed on the window
  │   → assert_true: If this is failing: DANGER, are you sure you want to expose the new interface Worklet to all webpages as a property on the global? Do not make a change to this file without review from jdm or Ms2ger for that specific change! expected true got false
  │ 
  │ test_interfaces/<@http://web-platform.test:8000/_mozilla/mozilla/interfaces.js:73:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  │ test@http://web-platform.test:8000/resources/testharness.js:497:9
  │ test_interfaces@http://web-platform.test:8000/_mozilla/mozilla/interfaces.js:2:3
  └ @http://web-platform.test:8000/_mozilla/mozilla/interfaces.html:13:1
```
We can either hide the Worklet interface behind an off-by-default pref or update the test to include the interface in the expected list.

**Comment by asajeffrey** [comment]
I added `Worklet` to the interfaces test. @bors-servo r=jdm

**Pull request merged**
